android/Drawer/app/src/main/java/com/example/drawer/MQTTController.java

```diff
@@ -1,109 +1,165 @@
+                public void messageArrived(String s, MqttMessage mqttMessage) {
+                    String message = new String(mqttMessage.getPayload());
+                    Log.d(SUBTAG, "Topic:   " + s);
+                    Log.d(SUBTAG, "Message: " + message);
+                    // TODO optimize, not crucial due to small size of hashmap, but needed eventually. -MH
+                    // potentially make it topic keyed, then have lists of objects, could be good
+                    subscriptionMap.forEach((text, topic) -> {
+                        Log.d(SUBTAG, ("text: " + text));
+                        if (s.equals(topic)) text.setText(message);
+                    });
+                }
+
+                @Override
+                public void deliveryComplete(IMqttDeliveryToken token) {
+                    Log.d(PUBTAG, "Delivery complete.");
+                }
+            });
         } catch (MqttException e) {
             //Standard error printing
-            System.out.println("Could not connectd");
-            System.out.println("reason " + e.getReasonCode());
-            System.out.println("msg " + e.getMessage());
-            System.out.println("loc " + e.getLocalizedMessage());
-            System.out.println("cause " + e.getCause());
-            System.out.println("excep " + e);
+            Log.d(ETAG, "Could not connect");
+            Log.d(ETAG, "reason " + e.getReasonCode());
+            Log.d(ETAG, "msg " + e.getMessage());
+            Log.d(ETAG, "loc " + e.getLocalizedMessage());
+            Log.d(ETAG, "cause " + e.getCause());
+            Log.d(ETAG, "except " + e);
             e.printStackTrace();
         }
     }
 
-    public static boolean isConnected() {
-        return mqttClient.isConnected();
+    public static boolean notConnected() {
+        if (mqttClient == null) return true;
+        return !mqttClient.isConnected();
+    }
+
+    public static void update(TextView text, String topic) {
+        subscriptionMap.put(text, topic);
     }
 
     public static void subscribe(String topic) {
-        if (!isConnected()) {
-            System.out.println("Not connected to MQTT broker.");
+        if (notConnected()) {
+            Log.d(ETAG, "Not connected to MQTT broker.");
             return;
         }
         try {
             mqttClient.subscribe(topic, 0);
-            System.out.println("Subscribed to: " + topic);
+            Log.d(SUBTAG, "Subscribed to: " + topic);
         } catch (MqttException e) {
             //Standard error printing
-            System.out.println("Subscription could not be performed");
-            System.out.println("reason " + e.getReasonCode());
-            System.out.println("msg " + e.getMessage());
-            System.out.println("loc " + e.getLocalizedMessage());
-            System.out.println("cause " + e.getCause());
-            System.out.println("excep " + e);
+            Log.d(ETAG, "Subscription could not be performed");
```

Comment:
Lines 107 - 112 are another example of excessive logging in my opinion. A similar case occurs towards the end of the publish() and disconnect() methods

Reply:
Fixed in most recent commit.